src/dmd/root/stringtable.d

```diff
@@ -70,48 +82,35 @@ pure:
 struct StringTable
 {
 private:
-    StringEntry* table;
-    size_t tabledim;
-    ubyte** pools;
-    size_t npools;
+    StringEntry[] table;
+    ubyte*[] pools;
     size_t nfill;
     size_t count;
     size_t countTrigger;   // amount which will trigger growing the table
 
-nothrow:
 public:
-    void _init(size_t size = 0) nothrow pure
+nothrow:
+    void _init(size_t size = 0) pure
     {
         size = nextpow2((size * loadFactorDenominator) / loadFactorNumerator);
         if (size < 32)
             size = 32;
-        table = cast(StringEntry*)mem.xcalloc(size, (table[0]).sizeof);
-        tabledim = size;
-        countTrigger = (tabledim * loadFactorNumerator) / loadFactorDenominator;
+        table = (cast(StringEntry*)mem.xcalloc(size, (table[0]).sizeof))[0 .. size];
+        countTrigger = (table.length * loadFactorNumerator) / loadFactorDenominator;
         pools = null;
-        npools = nfill = 0;
+        nfill = 0;
         count = 0;
     }
 
-    void reset(size_t size = 0) nothrow pure
```

Comment:
I'm not sure if DMD establishes whether or not it prefers global attributes vs local attributes (i.e. `nothrow:` vs `nothrow` on the function).  However, if there is no established style I would avoid making changes to it as it results in endless code churn.  You come in and remove the local ones, and then someone comes in later to adds them back in because they prefer the local ones where it's easier to tell when a function actually is nothrow.  It's an endless cycle.  Of course, if there is an established style then it breaks the cycle...is there one?  If not I would remove these parts of your change.

Reply:
It is established to prefer local attributes. Andrei and Walter prefer them because global attributes have resulted in a few accidental bugs in the past as they were simply overlooked.

Reply:
It's not super-explicit in its current form, but the DStyle says:

> Non-templated functions should be annotated with matching attributes (@nogc, @safe, pure, nothrow).

https://dlang.org/dstyle.html#phobos_attributes

Reply:
Thanks @wilzbach .  So in this case, let's keep the local `nothrow` attributes as they are.  If anything, it sounds like you could remove the global `nothrow:` attribute on line 92 if it can be removed.

There's not a document that describes these styles/decisions is there?  I remember Walter making a PR with a document describing general "code standards" but it was controversial.

Reply:
>  it sounds like you could remove the global nothrow: attribute on line 92 if it can be removed.

Yep.

> There's not a document that describes these styles/decisions is there? I remember Walter making a PR with a document describing general "code standards" but it was controversial.

It's here: https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md

Though it also says:

> Use attributes const/nothrow/pure/scope/@safe/private/etc. Successfully using pure functions is regarded with particular favor.

and:

> Reformatting into your personal style. Please stick with the existing style. Use the D Style for new code.

Reply:
I can minimize code churn and keep the diff as small as possible, or I can adopt D-style and do (some of) the following:
- make attributes local
- make attributes alphabetical. Note that before this PR, `lookup` was `inout nothrow pure @nogc` which would become `inout @nogc nothrow pure`. 
- annotate unittests. I don't know if that is needed for `nextPowerOf2`, it mentions "**Every** _unittest_ should be annotated ... on the templated function.". So does every unittest need it or every unittest that tests a templated function?

(Btw: ironic how D-style gives as an example for an annotates unittest "`pure nothrow @nogc @safe unittest { ... }`" which is not alphabetical)

Reply:
I would say, never make code style changes unless there is an established style.  This is the only way to avoid useless churn.  In this case, it sounds like local attributes are preferred so have at it.

In the case of "alphabetical" attributes, I don't know if there is an established style for that.  Sounds like a reasonable style to have to keep things consistent and easy to find, but just make sure it's an accepted style before making the change.